### PR TITLE
udp_tunneling: add udp tunneling

### PR DESCRIPTION
**Pull request by ohadvano**
Additional Description: Last bit of #29277. This PR uses the previous PR classes (#29733) to support raw UDP tunneling over HTTP stream, either with ``connect-udp`` upgrade or ``POST`` request.
Risk Level: low, guarded by config
Testing: unit tests, integration tests
Docs Changes: API, docs, config examples
Release Notes: none
Platform Specific Features: none

### CODE REVIEW

**Comment by ohadvano** [comment]
/assign @DavidSchinazi 
Thanks again for sticking with the reviews - this is the last PR for UDP tunneling. It adds the API config and uses the classes from previous PR to establish the tunnel, adds docs and integration testing.
Will appreciate your review, thanks

**Comment by ohadvano** [comment]
@mattklein123 would you have time to review the API changes alone? I assume Ryan may do a second pass afterwards, as he reviewed the previous PR

**Comment by mattklein123** [comment]
Yes I will do the API review tomorrow.

File: api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto [hunk 1]
```diff
@@ -62,6 +64,88 @@ message UdpProxyConfig {
...
+  // Configuration for tunneling UDP over other transports or application layers.
+  // Tunneling is currently supported over HTTP/2.
+  // [#next-free-field: 10]
+  message UdpTunnelingConfig {
+    // Configuration for UDP datagrams buffering.
+    message BufferOptions {
+      // If set, the filter will only buffer datagrams up to the requested limit, and will drop
+      // new UDP datagrams if the buffer contains the max_buffered_datagrams value at the time
+      // of a new datagram arrival. If not set, the default value is 1024 datagrams.
+      google.protobuf.UInt32Value max_buffered_datagrams = 1;
```
Comment (mattklein123) [inline api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto:76]: It seems like max bytes is always preferable to max datagrams so I'm not sure why someone would use this, but I don't feel strongly about it.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
In general API LGTM, a few small comments, thanks.

File: api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto [hunk 2]
```diff
@@ -62,6 +64,88 @@ message UdpProxyConfig {
...
+    //
+    bool use_post = 5;
+
+    // The path used with POST method. Default path is ``/``. If post path is specified and
+    // use_post field isn't true, it will be rejected.
+    string post_path = 6;
+
+    // The maximum number of unsuccessful connection attempts that will be made before giving up.
+    // If the parameter is not specified, 1 connection attempt will be made.
+    google.protobuf.UInt32Value max_connect_attempts = 7;
```
[the remark this answers is not in the document]
Comment (ohadvano) [reply api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto]: done, thanks

File: api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto [hunk 3]
```diff
@@ -62,6 +64,88 @@ message UdpProxyConfig {
...
+
+    // Additional request headers to upstream proxy. Neither ``:-prefixed`` pseudo-headers
+    // nor the Host: header can be overridden.
+    repeated config.core.v3.HeaderValueOption headers_to_add = 8
+        [(validate.rules).repeated = {max_items: 1000}];
```
[the remark this answers is not in the document]
Comment (ohadvano) [reply api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto]: I think you meant the use of command operators in the value? If yes, the answer is yes - added a comment and an example.

File: api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto (hunk 1 above)
Comment (ohadvano) [reply api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto:76]: The case this was meant to guard was when the downstream sends lots of small UDP datagrams (let's say a lot of 1 byte datagrams). Anyways, this is configurable so if the consumer doesn't need this kind of limit they can set it to max int value

**Review by mattklein123 — APPROVED** [review]
API LGTM, thanks.

File: api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto [hunk 4]
```diff
@@ -62,6 +64,104 @@ message UdpProxyConfig {
...
+    // the filter state key ``udp.connect.target_port``.
+    uint32 default_target_port = 4 [(validate.rules).uint32 = {lte: 65535 gt: 0}];
+
+    // Use POST method instead of CONNECT method to tunnel the UDP stream.
+    //
+    // .. note::
+    //   If use_post is set, the upstream stream does not comply with the connect-udp RFC, and
+    //   instead it will be a POST request. the path used in the headers will be set from the
+    //   post_path field, and the headers will not contain the target host and target port, as
+    //   required by the connect-udp protocol. This flag should be carefully used.
```
Comment (DavidSchinazi) [inline api/envoy/extensions/filters/udp/udp_proxy/v3/udp_proxy.proto:130]: nit:
```suggestion
    //   required by the connect-udp protocol. This flag should be used carefully.
```

File: docs/root/intro/arch_overview/http/upgrades.rst [hunk 5]
```diff
@@ -182,6 +182,52 @@ must be set to terminate ``CONNECT-UDP`` requests.
...
+.. note::
+   Raw UDP tunneling is in an alpha status and may not be stable enough for use in production.
+   We recommend to use this feature with caution.
+
+Apart from ``CONNECT-UDP`` termination, as described in the section above, Envoy also has support for tunneling raw UDP over HTTP
+``CONNECT`` or HTTP ``POST`` requests, by utilizing the UDP Proxy listener filter. By default, UDP tunneling is disabled, and can be
+enabled by setting the configuration for :ref:`tunneling_config <envoy_v3_api_field_extensions.filters.udp.udp_proxy.v3.UdpProxyConfig.tunneling_config>`.
+
+.. note::
+   Currently, Envoy only supports UDP tunneling over HTTP/2 streams.
```
Comment (DavidSchinazi) [inline docs/root/intro/arch_overview/http/upgrades.rst:199]: Where does this limitation come from?

File: docs/root/intro/arch_overview/http/upgrades.rst [hunk 6]
```diff
@@ -182,6 +182,52 @@ must be set to terminate ``CONNECT-UDP`` requests.
...
+
+By default, the ``tunneling_config`` will upgrade the connection to create HTTP/2 streams for each UDP session (a UDP session is identified by the datagrams 5-tuple), according
+to the `Proxying UDP in HTTP RFC <https://www.rfc-editor.org/rfc/rfc9298.html>`_. Since this upgrade protocol requires an encapsulation mechanism to preserve the boundaries of the original datagram,
+it's required to apply the :ref:`HTTP Capsule <config_udp_session_filters_http_capsule>` session filter.
+The HTTP/2 streams will be multiplexed over the upstream connection.
+
+As opposed to TCP tunneling, where downstream congestion can be applied by alternately disabling the read from the connection socket, for UDP datagrams, this mechanism is not supported.
```
Comment (DavidSchinazi) [inline docs/root/intro/arch_overview/http/upgrades.rst:206]:
```suggestion
As opposed to TCP tunneling, where downstream flow control can be applied by alternately disabling the read from the connection socket, for UDP datagrams, this mechanism is not supported.
```

File: docs/root/intro/arch_overview/http/upgrades.rst [hunk 7]
```diff
@@ -182,6 +182,52 @@ must be set to terminate ``CONNECT-UDP`` requests.
...
+Therefore, when tunneling UDP and a new datagram is received from the downstream, it is either streamed upstream, if the upstream is ready or halted by the UDP Proxy.
+In case the upstream is not ready (for example, when waiting for HTTP response headers), the datagram can either be dropped or buffered until the upstream is ready.
+In such cases, by default, downstream datagrams will be dropped, unless :ref:`buffer_options <envoy_v3_api_field_extensions.filters.udp.udp_proxy.v3.UdpProxyConfig.UdpTunnelingConfig.buffer_options>`
+is set by the ``tunneling_config``. The default buffer limits are modest to try and prevent a lot of unwanted buffered memory, but can and should be adjusted per the required use-case.
+When the upstream becomes ready, the UDP Proxy will first flush all the previously buffered datagrams.
+
+.. note::
+   If ``POST`` is set, the upstream stream does not comply with the connect-udp RFC, and instead it will be a POST request.
+   The path used in the headers will be set from the post_path field, and the headers will not contain the target host and
+   target port, as required by the connect-udp protocol. This option should be carefully used.
```
Comment (DavidSchinazi) [inline docs/root/intro/arch_overview/http/upgrades.rst:216]: nit
```suggestion
   target port, as required by the connect-udp protocol. This option should be used carefully.
```

File: source/common/http/header_utility.cc [hunk 8]
```diff
@@ -239,7 +239,7 @@ bool HeaderUtility::isConnect(const RequestHeaderMap& headers) {
   return headers.Method() && headers.Method()->value() == Http::Headers::get().MethodValues.Connect;
 }
 
-bool HeaderUtility::isConnectUdp(const RequestHeaderMap& headers) {
+bool HeaderUtility::isConnectUdp(const RequestOrResponseHeaderMap& headers) {
```
Comment (DavidSchinazi) [inline source/common/http/header_utility.cc:242]: Why this change? Response headers don't indicate connect-udp

File: source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h [hunk 9]
```diff
@@ -294,7 +301,14 @@ class HttpUpstreamImpl : public HttpUpstream, protected Http::StreamCallbacks {
     }
 
     void decodeHeaders(Http::ResponseHeaderMapPtr&& headers, bool end_stream) override {
-      bool is_valid_response = Http::CodeUtility::is2xx(Http::Utility::getResponseStatus(*headers));
+      auto status = Http::Utility::getResponseStatus(*headers);
+      bool is_valid_post = parent_.tunnel_config_.usePost() && Http::CodeUtility::is2xx(status);
+      // In connect-udp case, Envoy will transform the H2 headers to H1 upgrade headers.
```
Comment (DavidSchinazi) [inline source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h:306]: I knew Envoy did this for requests, but does it do it for responses too? Can you point me at the code that does that?

File: source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h [hunk 10]
```diff
@@ -294,7 +301,14 @@ class HttpUpstreamImpl : public HttpUpstream, protected Http::StreamCallbacks {
...
+      // A valid response should have SwitchingProtocol status and connect-udp upgrade.
+      bool is_valid_upgrade = status == 101 && Http::HeaderUtility::isConnectUdp(*headers);
+
+      ASSERT(!(is_valid_post && is_valid_upgrade));
+      bool is_valid_response = is_valid_post || is_valid_upgrade;
```
Comment (DavidSchinazi) [inline source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h:311]: Shouldn't we look at the config to see if we're in POST-mode instead? This will incorrectly accept some responses otherwise

**Review by DavidSchinazi — COMMENTED** [review]
Nice!

File: docs/root/intro/arch_overview/http/upgrades.rst (hunk 5 above)
Comment (ohadvano) [reply docs/root/intro/arch_overview/http/upgrades.rst:199]: Not a technical implementation, only since I focused on ``HTTP/2`` implementation for this round. I intend to add HTTP/3 support and tests along the way.

File: source/common/http/header_utility.cc (hunk 8 above)
Comment (ohadvano) [reply source/common/http/header_utility.cc:242]: Envoy will add the connect-udp upgrade header, see other comment for elaborate code explanation

File: source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h (hunk 9 above)
Comment (ohadvano) [reply source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h:306]: Yeah so in the request path, if the headers are an upgrade, Envoy will expect them to be in ``H1`` form, and will transform to ``H2``, and save the upgrade type in a member, see here:
https://github.com/envoyproxy/envoy/blob/7479b6cbdde02186b2fdbad3137965f628d3f42f/source/common/http/http2/codec_impl.cc#L278

On the response path, if the upgrade was set, Envoy will transform the ``H2`` headers to ``H1`` and set the upgrade as a header, see:
https://github.com/envoyproxy/envoy/blob/7479b6cbdde02186b2fdbad3137965f628d3f42f/source/common/http/http2/codec_impl.cc#L534

File: source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h (hunk 10 above)
Comment (ohadvano) [reply source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h:311]: ``is_valid_post`` can be true only if the config is set to ``usePost``, see line 305 above.
``is_valid_post`` and ``is_valid_upgrade`` can never be both true (asserted by line 310).
Not sure if I fully understood your comment, please lmk otherwise.

File: docs/root/intro/arch_overview/http/upgrades.rst (hunk 5 above)
Comment (DavidSchinazi) [reply docs/root/intro/arch_overview/http/upgrades.rst:199]: Sounds good. Can we say here that h3 is coming soon, or add a TODO?

File: source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h (hunk 9 above)
Comment (DavidSchinazi) [reply source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h:306]: Thanks!

File: source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h (hunk 10 above)
Comment (DavidSchinazi) [reply source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h:311]: Ah, you're right. Thanks for pointing it out. Could I propose to slightly refactor the code to make it easier to read? Perhaps:

```c++
      auto status = Http::Utility::getResponseStatus(*headers);
      bool is_valid_response;
      if (parent_.tunnel_config_.usePost()) {
        is_valid_response = Http::CodeUtility::is2xx(status);
      } else {
        // In connect-udp case, Envoy will transform the H2 headers to H1 upgrade headers.
        // A valid response should have SwitchingProtocol status and connect-udp upgrade.
        is_valid_response = status == 101 && Http::HeaderUtility::isConnectUdp(*headers);
      }
```

**Review by DavidSchinazi — COMMENTED** [review]
Thanks!

File: source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h [hunk 11]
```diff
@@ -294,7 +301,14 @@ class HttpUpstreamImpl : public HttpUpstream, protected Http::StreamCallbacks {
     }
 
     void decodeHeaders(Http::ResponseHeaderMapPtr&& headers, bool end_stream) override {
-      bool is_valid_response = Http::CodeUtility::is2xx(Http::Utility::getResponseStatus(*headers));
+      auto status = Http::Utility::getResponseStatus(*headers);
+      bool is_valid_post = parent_.tunnel_config_.usePost() && Http::CodeUtility::is2xx(status);
+      // In connect-udp case, Envoy will transform the H2 headers to H1 upgrade headers.
+      // A valid response should have SwitchingProtocol status and connect-udp upgrade.
+      bool is_valid_upgrade = status == 101 && Http::HeaderUtility::isConnectUdp(*headers);
```
Comment (DavidSchinazi) [inline source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h:308]: Can we move the `status == 101` check to inside `Http::HeaderUtility::isConnectUdp()`?

File: source/common/http/header_utility.h [hunk 12]
```diff
@@ -167,7 +167,7 @@ class HeaderUtility {
   /**
    * @brief a helper function to determine if the headers represent a CONNECT-UDP request.
    */
-  static bool isConnectUdp(const RequestHeaderMap& headers);
+  static bool isConnectUdp(const RequestOrResponseHeaderMap& headers);
```
Comment (DavidSchinazi) [inline source/common/http/header_utility.h:170]: Instead of merging them, can we instead have two methods `isConnectUdpRequest()` and `isConnectUdpResponse()`? That's less likely to cause security issues due to confusion

File: docs/root/intro/arch_overview/http/upgrades.rst (hunk 5 above)
Comment (ohadvano) [reply docs/root/intro/arch_overview/http/upgrades.rst:199]: There's a TODO in line 734 (``// TODO(ohadvano): support upstream HTTP/3.``). I don't think writing in the docs themselves that H3 is coming soon would be the correct action

File: source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h (hunk 10 above)
Comment (ohadvano) [reply source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h]: Yeah looks better, thanks

**Review by DavidSchinazi — APPROVED** [review]
Thanks!

**Comment by ohadvano** [comment]
Thanks for your review @DavidSchinazi.

/assign @RyanTheOptimist 
Since you reviewed the previous PR, I'll appreciate if you could review this one as well, as they are tightly related. Thanks. I assumed you'd be the right person to review this, but if it doesn't work for you or you don't currently have time to review, please assign someone else.

@mattklein123 not sure it's mandatory now, but I made a small rephrase in the API due to David's comment, can you please stamp it again?

File: source/extensions/filters/udp/udp_proxy/hash_policy_impl.cc [hunk 13]
```diff
@@ -48,7 +50,7 @@ HashPolicyImpl::HashPolicyImpl(
     hash_impl_ = std::make_unique<KeyHashMethod>(hash_policies[0]->key());
     break;
   case UdpProxyConfig::HashPolicy::PolicySpecifierCase::POLICY_SPECIFIER_NOT_SET:
-    PANIC_DUE_TO_CORRUPT_ENUM;
+    throw EnvoyException("policy specifier not set");
```
Comment (RyanTheOptimist) [inline source/extensions/filters/udp/udp_proxy/hash_policy_impl.cc:53]: Why this change?

File: source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h [hunk 14]
```diff
@@ -612,6 +625,53 @@ class UdpProxyFilter : public Network::UdpListenerReadFilter,
     const bool use_original_src_ip_;
   };
 
+  class TunnelingActiveSession : public ActiveSession,
```
Comment (RyanTheOptimist) [inline source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h:628]: nit: brief class comment.

File: test/extensions/filters/udp/udp_proxy/udp_proxy_filter_test.cc [hunk 15]
```diff
@@ -1468,10 +1492,13 @@ class HttpUpstreamImplTest : public testing::Test {
...
                   absl::optional<std::string> opt_path = absl::nullopt,
                   absl::optional<HeaderToAdd> header_to_add = absl::nullopt) {
+    // In case connect-udp is used, Envoy expect the H2 headers to be normalized with H1,
+    // so expect that the request headers here match H1 CONNECT headers, even though
+    // eventually H2 headers will be sent.
     std::string scheme = is_ssl ? "https" : "http";
     std::string authority =
         opt_authority.has_value() ? opt_authority.value() : "default.host.com:10";
-    std::string method = use_post ? "POST" : "CONNECT";
+    std::string method = use_post ? "POST" : "GET";
```
Comment (RyanTheOptimist) [inline test/extensions/filters/udp/udp_proxy/udp_proxy_filter_test.cc:1501]: The comment just above says to expect H1 CONNECT but here we use "GET". This is surprising. Can you elaborate?

**Review by RyanTheOptimist — COMMENTED** [review]
Looking great!

File: test/extensions/filters/udp/udp_proxy/udp_proxy_filter_test.cc (hunk 15 above)
Comment (ohadvano) [reply test/extensions/filters/udp/udp_proxy/udp_proxy_filter_test.cc:1503]: Done

File: source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h (hunk 14 above)
Comment (ohadvano) [reply source/extensions/filters/udp/udp_proxy/udp_proxy_filter.h:633]: Done

File: source/extensions/filters/udp/udp_proxy/hash_policy_impl.cc (hunk 13 above)
Comment (ohadvano) [reply source/extensions/filters/udp/udp_proxy/hash_policy_impl.cc:53]: I was trying to increase coverage for the UDP folder as a whole, and apparently, before this change this line (the PANIC) was not covered, even though there was a test that tests this scenario. I wasn't able to understand why even though a test checks this, the coverage tool considers this line as not covered, but changing to exception instead makes the trick.

Anyways, this isn't 100% related to this PR, I agree - but since it's a super small change I was hoping to piggyback this PR instead of opening a new one.

Comment (RyanTheOptimist) [reply source/extensions/filters/udp/udp_proxy/hash_policy_impl.cc:53]: We're actually trying to remove exceptions from the codebase. See #29840. So adding a new exception would be a step backward from that objective. Let's revert this part of the change.

Comment (ohadvano) [reply source/extensions/filters/udp/udp_proxy/hash_policy_impl.cc]: Done

**Review by RyanTheOptimist — DISMISSED** [review]

**Comment by ohadvano** [comment]
@RyanTheOptimist I made a small refactor, can you take a look again please? Thanks.

**Review by RyanTheOptimist — APPROVED** [review]

**Review by phlax — APPROVED** [review]
verify/examples failure is unrelated - fixed in #30002 

lgtm, thanks @ohadvano
